src/stdlib/replace_with.rs

```diff
@@ -0,0 +1,248 @@
+use crate::value;
+
+fn replace_with<T>(
+    value: Value,
+    pattern: &Regex,
+    count: Value,
+    ctx: &mut Context,
+    runner: closure::Runner<T>,
+) -> Resolved
+where
+    T: Fn(&mut Context) -> Result<Value, ExpressionError>,
+{
+    let haystack = value.try_bytes_utf8_lossy()?;
+    let count = match count.try_integer()? {
+        i if i > 0 => i as usize,
+        i if i < 0 => 0,
+        // this is when i == 0
+        _ => return Ok(value),
+    };
+    let captures = pattern.captures_iter(&haystack);
+    make_replacement(
+        captures,
+        &haystack,
+        count,
+        pattern.capture_names(),
+        ctx,
+        runner,
+    )
+}
+
+fn make_replacement<T>(
+    caps: CaptureMatches,
+    haystack: &str,
+    count: usize,
+    capture_names: CaptureNames,
+    ctx: &mut Context,
+    runner: closure::Runner<T>,
+) -> Resolved
+where
+    T: Fn(&mut Context) -> Result<Value, ExpressionError>,
+{
+    // possible optimization: peek at first capture, if none return the original value.
+    let mut replaced = String::with_capacity(haystack.len());
+    let limit = if count == 0 { usize::MAX } else { count - 1 };
+    let mut last_match = 0;
+    // we loop over the matches ourselves instead of calling Regex::replacen, so that we can
+    // handle errors. This is however based on the implementation of Regex::replacen
+    for (idx, captures) in caps.enumerate() {
+        // Safe to unrap because the 0th index always includes the full match.
+        let m = captures.get(0).unwrap(); // full match
+
+        let mut value = captures_to_value(&captures, capture_names.clone());
+        runner.map_value(ctx, &mut value)?;
+        let replacement = value.try_bytes_utf8_lossy()?;
+
+        replaced.push_str(&haystack[last_match..m.start()]);
+        replaced.push_str(&replacement);
+        last_match = m.end();
+        if idx >= limit {
+            break;
+        }
+    }
+    // add the final component
+    replaced.push_str(&haystack[last_match..]);
+    Ok(replaced.into())
+}
+
+fn captures_to_value(captures: &Captures, capture_names: CaptureNames) -> Value {
+    let full_match: Value = captures.get(0).unwrap().as_str().into();
+    // The length includes the total match, so subtract 1
+    let mut capture_groups: Vec<Value> = Vec::with_capacity(captures.len() - 1);
+    let mut named_groups: ObjectMap = BTreeMap::new();
+
+    // We skip the first entry, because it is for the full match, which we have already
+    // extracted
+    for (idx, name) in capture_names.enumerate().skip(1) {
+        let value: Value = if let Some(group) = captures.get(idx) {
+            group.as_str().into()
+        } else {
+            Value::Null
+        };
+        if let Some(name) = name {
+            named_groups.insert(name.into(), value.clone());
+        }
+        capture_groups.push(value);
+    }
+
+    value!({
+        string: full_match,
+        captures: capture_groups,
+        named: named_groups,
```

Comment:
I wonder if it would be better to put the named capture groups directly on the the returned object.  But then what do we do if there is a capture group named "string" or "captures"?

Reply:
This sounds cleaner. Not sure I understand the question. Can you give an example?

Reply:
If the regex is something like `r'"(?P<string>.*)"''` should I:

1. Set the `string` property to the capture group for "string"?
2. Rename the capture group to something else like "string_"
3. Not use a named property for the capture group
4. Return an error because there is an illegal name

Reply:
Ah makes sense, thanks for the clear explanation!

Let's go for 4.